### PR TITLE
[ENG-968] use filecache when getting custom field config from compare() method

### PR DESCRIPTION
**Pull request by scottshipman**
**Please be sure you are submitting this against the _master_ branch.**

[//]: # This Pull Request (Place an 'X' for each):

| Risk Level                                | No | Low | High |
| ----------------------------------------- | -- | --- | ---- |
| Alters Lead Data?                         |  X  |     |      |
| Schema Change?                            | X   |     |      |
| Adds A Query or Modifies Existing Query?  |    | X    |      |
| Adds or Modifies Existing Auto-Enhancer?  | X   |     |      |
| Modifies Ingestion Process?               | X   |     |      |
| Modifies sendContact Data?                |  X  |     |      |

[//]: # ( Required: )

### CODE REVIEW

File: Entity/OverrideLeadFieldRepository.php [hunk 1]
```diff
@@ -255,7 +278,7 @@ public function getValueList($field, $search = '', $limit = 10, $start = 0)
         $fieldModel = $this->fieldModel;
 
         // get list of extendedFields
-        if ($extendedField = $this->getExtendedField($field)) {
+        if ($extendedField = $this->extendedFieldConfigurations[$field]) {
```
Comment (heathdutton) [inline Entity/OverrideLeadFieldRepository.php]: Won't this throw a warning if not set?

File: Entity/OverrideLeadFieldRepository.php [hunk 2]
```diff
@@ -32,10 +40,12 @@ class OverrideLeadFieldRepository extends LeadFieldRepository
...
      * @param ExtendedFieldModel $fieldModel
      */
-    public function __construct(EntityManager $em, ClassMetadata $class, ExtendedFieldModel $fieldModel)
+    public function __construct(EntityManager $em, ClassMetadata $class, ExtendedFieldModel $fieldModel, CoreParametersHelper $coreParametersHelper)
     {
         parent::__construct($em, $class);
-        $this->fieldModel = $fieldModel;
+        $this->fieldModel                  = $fieldModel;
+        $this->coreParametersHelper        = $coreParametersHelper;
+        $this->extendedFieldConfigurations = $this->getExtendedFieldConfigurations();
```
Comment (heathdutton) [inline Entity/OverrideLeadFieldRepository.php]: I don't think we should ever put something that requires a query in a constructor. Rather have it populate as needed (and cache). Not all mautic actions require this array.

File: Entity/OverrideLeadFieldRepository.php [hunk 3]
```diff
@@ -212,29 +221,46 @@ public function compareValue($lead, $field, $value, $operatorExpr)
...
-            ->setParameter('alias', $alias);
+        if (!$this->extendedFieldConfigurations) {
+            $cacheHelper     = new CacheStorageHelper(
+                CacheStorageHelper::ADAPTOR_FILESYSTEM,
+                'extendedFieldConfiguration',
+                null,
+                $this->coreParametersHelper->getParameter('mautic.cache_path'),
+                60
+            );
+            if (empty($extendedFieldConfigurations = $cacheHelper->get('extendedFieldConfiguration', 60))) {
```
Comment (dean-martin) [inline Entity/OverrideLeadFieldRepository.php]: I've seen this before, but I don't really understand it.
`empty($extendedFieldConfigurations = $cacheHelper->get('extendedFieldConfiguration', 60))`
It's setting `$extendedFieldConfigurations` to the cached version, and then evaluating if $extendedFieldConfigurations is not null?

File: Entity/OverrideLeadFieldRepository.php [hunk 4]
```diff
@@ -212,29 +221,46 @@ public function compareValue($lead, $field, $value, $operatorExpr)
...
-            )
-            ->setParameter('alias', $alias);
+        if (!$this->extendedFieldConfigurations) {
+            $cacheHelper     = new CacheStorageHelper(
+                CacheStorageHelper::ADAPTOR_FILESYSTEM,
+                'extendedFieldConfiguration',
+                null,
+                $this->coreParametersHelper->getParameter('mautic.cache_path'),
+                60
+            );
```
Comment (dean-martin) [inline Entity/OverrideLeadFieldRepository.php]: Ideally I'd prefer if CacheStorageHelper was passed through the constructor, so it's more configurable to the consumer of the class.

File: Entity/OverrideLeadFieldRepository.php (hunk 3 above)
Comment (heathdutton) [reply Entity/OverrideLeadFieldRepository.php]: Yes, that's what he's doing here.
```
php > var_dump(boolval(empty($a = null)));
bool(true)
php > var_dump(boolval(empty($a = 1)));
bool(false)
```

The `empty` in this case isn't technically required, but scrutinizer might demand it.

File: Entity/OverrideLeadFieldRepository.php (hunk 4 above)
Comment (heathdutton) [reply Entity/OverrideLeadFieldRepository.php]: Normally, yes. But this is an override class, and it'll never be consumed in such a way. Best to keep the constructor as stock as possible to reduce collisions when core changes.

File: Entity/OverrideLeadFieldRepository.php (hunk 3 above)
Comment (scottshipman) [reply Entity/OverrideLeadFieldRepository.php]: not necessarily null, could be an empty array. Not likely in a system that actually has fields. but technically, cache is set to be an array of custom field configs, and no custom fields would be an empty array. if there is no instance of the cache key, then null may be returned.

**Pull request merged**